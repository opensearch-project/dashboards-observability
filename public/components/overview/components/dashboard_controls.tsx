/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiButtonIcon,
  EuiFlexGroup,
  EuiFlexItem,
  EuiIcon,
  EuiLink,
  EuiSuperDatePicker,
  EuiText,
  EuiToolTip,
} from '@elastic/eui';
import { OnTimeChangeProps } from '@opensearch-project/oui/src/eui_components/date_picker/super_date_picker/super_date_picker';
import React from 'react';
import { useObservable } from 'react-use';
import { FormattedMessage } from '@osd/i18n/react';
import { coreRefs } from '../../../framework/core_refs';
import { HOME_CONTENT_AREAS } from '../../../plugin_helpers/plugin_overview';
import { redirectToDashboards } from '../../getting_started/components/utils';
import { AddDashboardCallout } from './add_dashboard_callout';
import { ObsDashboardStateManager } from './obs_dashboard_state_manager';

export function DashboardControls() {
  const isDashboardSelected = useObservable(ObsDashboardStateManager.isDashboardSelected$);
  const dashboardState = useObservable(ObsDashboardStateManager.dashboardState$);
  const showFlyout = useObservable(ObsDashboardStateManager.showFlyout$);

  const onTimeChange = (onTimeChangeProps: OnTimeChangeProps) => {
    ObsDashboardStateManager.dashboardState$.next({
      ...dashboardState!,
      startDate: onTimeChangeProps.start,
      endDate: onTimeChangeProps.end,
    });

    coreRefs.contentManagement?.updatePageSection(HOME_CONTENT_AREAS.DASHBOARD, (section) => {
      if (section && section.kind === 'dashboard') {
        return {
          ...section,
          input: {
            ...section.input,
            timeRange: { to: onTimeChangeProps.end, from: onTimeChangeProps.start },
          },
        };
      }
      return section;
    });
  };

  return isDashboardSelected ? (
    <EuiFlexGroup gutterSize="s" alignItems="center" justifyContent="spaceBetween">
      <EuiFlexItem grow={false}>
        <EuiFlexGroup alignItems="center" gutterSize="s">
          <EuiFlexItem grow={false}>
            <EuiText size="m" className="obsOverviewDashboardHeader">
              <h4>{dashboardState?.dashboardTitle}</h4>
            </EuiText>
          </EuiFlexItem>
          <EuiFlexItem grow={false}>
            <EuiToolTip
              content={
                <FormattedMessage
                  id="observability.dashboard.popout.tooltip"
                  defaultMessage="Go to dashboard"
                />
              }
            >
              <EuiLink
                onClick={() => redirectToDashboards('/view/' + dashboardState?.dashboardId)}
                external={true}
              >
                <EuiIcon type="popout" />
              </EuiLink>
            </EuiToolTip>
          </EuiFlexItem>
        </EuiFlexGroup>
      </EuiFlexItem>
      <EuiFlexItem grow={false}>
        <EuiFlexGroup justifyContent="flexEnd" alignItems="center" gutterSize="s">
          <EuiFlexItem grow={false}>
            <EuiSuperDatePicker
              start={dashboardState?.startDate}
              end={dashboardState?.endDate}
              onTimeChange={onTimeChange}
              compressed
            />
          </EuiFlexItem>
          <EuiFlexItem grow={false}>
            <EuiToolTip content="Replace dashboard">
              <EuiButtonIcon
                iconType="gear"
                aria-label="Dashboard"
                color="primary"
                onClick={showFlyout}
                display="base"
                size="s"
              />
            </EuiToolTip>
          </EuiFlexItem>
        </EuiFlexGroup>
      </EuiFlexItem>
    </EuiFlexGroup>
  ) : (
    <AddDashboardCallout />
  );
}
