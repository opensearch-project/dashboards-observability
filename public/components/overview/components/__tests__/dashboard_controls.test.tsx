/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { DashboardControls } from '../dashboard_controls';

configure({ adapter: new Adapter() });

const mountDashboardControls = () => {
  return mount(<DashboardControls />);
};

jest.mock('../../../getting_started/components/utils', () => ({
  redirectToDashboards: jest.fn(),
}));

jest.mock('../../../../framework/core_refs', () => ({
  coreRefs: {
    contentManagement: {
      updatePageSection: jest.fn(),
    },
  },
}));

describe('Dashboard controls', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render', () => {
    const wrapper = mountDashboardControls();
    expect(wrapper).toMatchSnapshot();
  });
});
