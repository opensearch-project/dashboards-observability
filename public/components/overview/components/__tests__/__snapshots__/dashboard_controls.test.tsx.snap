// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard controls should render 1`] = `
<DashboardControls>
  <AddDashboardCallout>
    <EuiPanel
      hasShadow={true}
      paddingSize="m"
    >
      <div
        className="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
      >
        <EuiFlexGroup
          alignItems="center"
          direction="column"
          gutterSize="m"
          justifyContent="center"
        >
          <div
            className="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentCenter euiFlexGroup--directionColumn euiFlexGroup--responsive"
          >
            <EuiFlexItem
              grow={false}
            >
              <div
                className="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <EuiIcon
                  size="xxl"
                  type="gear"
                >
                  <EuiIconBeaker
                    aria-hidden={true}
                    className="euiIcon euiIcon--xxLarge euiIcon-isLoading"
                    focusable="false"
                    role="img"
                    style={null}
                  >
                    <svg
                      aria-hidden={true}
                      className="euiIcon euiIcon--xxLarge euiIcon-isLoading"
                      focusable="false"
                      height={16}
                      role="img"
                      style={null}
                      viewBox="0 0 16 16"
                      width={16}
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                      />
                    </svg>
                  </EuiIconBeaker>
                </EuiIcon>
              </div>
            </EuiFlexItem>
            <EuiFlexItem
              grow={false}
            >
              <div
                className="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <EuiTitle
                  size="s"
                >
                  <h3
                    className="euiTitle euiTitle--small"
                  >
                    Customize this page
                  </h3>
                </EuiTitle>
              </div>
            </EuiFlexItem>
          </div>
        </EuiFlexGroup>
        <EuiSpacer
          size="s"
        >
          <div
            className="euiSpacer euiSpacer--s"
          />
        </EuiSpacer>
        <EuiFlexGroup
          alignItems="stretch"
          gutterSize="m"
          justifyContent="center"
          wrap={true}
        >
          <div
            className="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--justifyContentCenter euiFlexGroup--directionRow euiFlexGroup--responsive euiFlexGroup--wrap"
          >
            <EuiFlexItem
              grow={false}
              style={
                Object {
                  "maxWidth": 300,
                }
              }
            >
              <div
                className="euiFlexItem euiFlexItem--flexGrowZero"
                style={
                  Object {
                    "maxWidth": 300,
                  }
                }
              >
                <EuiCard
                  description="Customize the overview page by adding a dashboard."
                  footer={
                    <EuiFlexGroup
                      gutterSize="none"
                      justifyContent="flexEnd"
                    >
                      <EuiSmallButton
                        color="primary"
                        fill={true}
                      >
                        Select a dashboard
                      </EuiSmallButton>
                    </EuiFlexGroup>
                  }
                  image={
                    <EuiImage
                      alt="Add a dashboard image"
                      size="m"
                      src="file-stub"
                      style={
                        Object {
                          "height": "150px",
                          "objectFit": "cover",
                          "width": "100%",
                        }
                      }
                    />
                  }
                  title="Add a dashboard"
                >
                  <EuiPanel
                    className="euiCard euiCard--centerAligned"
                    element="div"
                    hasShadow={true}
                  >
                    <div
                      className="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow euiCard euiCard--centerAligned"
                    >
                      <div
                        className="euiCard__top"
                      >
                        <div
                          className="euiCard__image"
                        >
                          <EuiImage
                            alt="Add a dashboard image"
                            size="m"
                            src="file-stub"
                            style={
                              Object {
                                "height": "150px",
                                "objectFit": "cover",
                                "width": "100%",
                              }
                            }
                          >
                            <figure
                              className="euiImage euiImage--medium"
                            >
                              <img
                                alt="Add a dashboard image"
                                className="euiImage__img"
                                src="file-stub"
                                style={
                                  Object {
                                    "height": "150px",
                                    "objectFit": "cover",
                                    "width": "100%",
                                  }
                                }
                              />
                            </figure>
                          </EuiImage>
                        </div>
                      </div>
                      <div
                        className="euiCard__content"
                      >
                        <EuiTitle
                          className="euiCard__title"
                          id="random_html_idTitle"
                          size="s"
                        >
                          <span
                            className="euiTitle euiTitle--small euiCard__title"
                            id="random_html_idTitle"
                          >
                            Add a dashboard
                          </span>
                        </EuiTitle>
                        <EuiText
                          className="euiCard__description"
                          id="random_html_idDescription"
                          size="s"
                        >
                          <div
                            className="euiText euiText--small euiCard__description"
                            id="random_html_idDescription"
                          >
                            <p>
                              Customize the overview page by adding a dashboard.
                            </p>
                          </div>
                        </EuiText>
                      </div>
                      <div
                        className="euiCard__footer"
                      >
                        <EuiFlexGroup
                          gutterSize="none"
                          justifyContent="flexEnd"
                        >
                          <div
                            className="euiFlexGroup euiFlexGroup--justifyContentFlexEnd euiFlexGroup--directionRow euiFlexGroup--responsive"
                          >
                            <EuiSmallButton
                              color="primary"
                              fill={true}
                            >
                              <EuiButton
                                color="primary"
                                fill={true}
                                size="s"
                              >
                                <EuiButtonDisplay
                                  baseClassName="euiButton"
                                  color="primary"
                                  disabled={false}
                                  element="button"
                                  fill={true}
                                  isDisabled={false}
                                  size="s"
                                  type="button"
                                >
                                  <button
                                    className="euiButton euiButton--primary euiButton--small euiButton--fill"
                                    disabled={false}
                                    style={
                                      Object {
                                        "minWidth": undefined,
                                      }
                                    }
                                    type="button"
                                  >
                                    <EuiButtonContent
                                      className="euiButton__content"
                                      iconGap="m"
                                      iconSide="left"
                                      textProps={
                                        Object {
                                          "className": "euiButton__text",
                                        }
                                      }
                                    >
                                      <span
                                        className="euiButtonContent euiButton__content"
                                      >
                                        <span
                                          className="euiButton__text"
                                        >
                                          Select a dashboard
                                        </span>
                                      </span>
                                    </EuiButtonContent>
                                  </button>
                                </EuiButtonDisplay>
                              </EuiButton>
                            </EuiSmallButton>
                          </div>
                        </EuiFlexGroup>
                      </div>
                    </div>
                  </EuiPanel>
                </EuiCard>
              </div>
            </EuiFlexItem>
            <EuiFlexItem
              grow={false}
              style={
                Object {
                  "maxWidth": 300,
                }
              }
            >
              <div
                className="euiFlexItem euiFlexItem--flexGrowZero"
                style={
                  Object {
                    "maxWidth": 300,
                  }
                }
              >
                <EuiCard
                  description="Log, Traces, and Metrics for an e-commerce application in OpenTelemetry standard."
                  footer={
                    <EuiFlexGroup
                      gutterSize="none"
                      justifyContent="flexEnd"
                    >
                      <EuiSmallButton
                        onClick={[Function]}
                      >
                        Add sample data
                      </EuiSmallButton>
                    </EuiFlexGroup>
                  }
                  image={
                    <EuiImage
                      alt="Install sample Observability data image"
                      size="m"
                      src="file-stub"
                      style={
                        Object {
                          "objectFit": "cover",
                          "width": "100%",
                        }
                      }
                    />
                  }
                  title="Install sample Observability data"
                >
                  <EuiPanel
                    className="euiCard euiCard--centerAligned"
                    element="div"
                    hasShadow={true}
                  >
                    <div
                      className="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow euiCard euiCard--centerAligned"
                    >
                      <div
                        className="euiCard__top"
                      >
                        <div
                          className="euiCard__image"
                        >
                          <EuiImage
                            alt="Install sample Observability data image"
                            size="m"
                            src="file-stub"
                            style={
                              Object {
                                "objectFit": "cover",
                                "width": "100%",
                              }
                            }
                          >
                            <figure
                              className="euiImage euiImage--medium"
                            >
                              <img
                                alt="Install sample Observability data image"
                                className="euiImage__img"
                                src="file-stub"
                                style={
                                  Object {
                                    "objectFit": "cover",
                                    "width": "100%",
                                  }
                                }
                              />
                            </figure>
                          </EuiImage>
                        </div>
                      </div>
                      <div
                        className="euiCard__content"
                      >
                        <EuiTitle
                          className="euiCard__title"
                          id="random_html_idTitle"
                          size="s"
                        >
                          <span
                            className="euiTitle euiTitle--small euiCard__title"
                            id="random_html_idTitle"
                          >
                            Install sample Observability data
                          </span>
                        </EuiTitle>
                        <EuiText
                          className="euiCard__description"
                          id="random_html_idDescription"
                          size="s"
                        >
                          <div
                            className="euiText euiText--small euiCard__description"
                            id="random_html_idDescription"
                          >
                            <p>
                              Log, Traces, and Metrics for an e-commerce application in OpenTelemetry standard.
                            </p>
                          </div>
                        </EuiText>
                      </div>
                      <div
                        className="euiCard__footer"
                      >
                        <EuiFlexGroup
                          gutterSize="none"
                          justifyContent="flexEnd"
                        >
                          <div
                            className="euiFlexGroup euiFlexGroup--justifyContentFlexEnd euiFlexGroup--directionRow euiFlexGroup--responsive"
                          >
                            <EuiSmallButton
                              onClick={[Function]}
                            >
                              <EuiButton
                                onClick={[Function]}
                                size="s"
                              >
                                <EuiButtonDisplay
                                  baseClassName="euiButton"
                                  disabled={false}
                                  element="button"
                                  isDisabled={false}
                                  onClick={[Function]}
                                  size="s"
                                  type="button"
                                >
                                  <button
                                    className="euiButton euiButton--primary euiButton--small"
                                    disabled={false}
                                    onClick={[Function]}
                                    style={
                                      Object {
                                        "minWidth": undefined,
                                      }
                                    }
                                    type="button"
                                  >
                                    <EuiButtonContent
                                      className="euiButton__content"
                                      iconGap="m"
                                      iconSide="left"
                                      textProps={
                                        Object {
                                          "className": "euiButton__text",
                                        }
                                      }
                                    >
                                      <span
                                        className="euiButtonContent euiButton__content"
                                      >
                                        <span
                                          className="euiButton__text"
                                        >
                                          Add sample data
                                        </span>
                                      </span>
                                    </EuiButtonContent>
                                  </button>
                                </EuiButtonDisplay>
                              </EuiButton>
                            </EuiSmallButton>
                          </div>
                        </EuiFlexGroup>
                      </div>
                    </div>
                  </EuiPanel>
                </EuiCard>
              </div>
            </EuiFlexItem>
          </div>
        </EuiFlexGroup>
      </div>
    </EuiPanel>
  </AddDashboardCallout>
</DashboardControls>
`;
